### PR TITLE
Add user locale/region setting

### PR DESCRIPTION
**Pull request by tcitworld**
This PR allows having a dedicated user locale/region setting. See https://github.com/nextcloud/server/issues/1781

Needs :
- [x] Where to put the list of locales (obviously this locales.json is going to be moved)
- [x] More tests (please point out where)
- [x] UI fixes on personal settings page
- [x] Provide a system wide setting for locale
- [x] Add a check on OCS endpoint for valid locale format
- [x] Pass locale to MomentJS
- [x] Don't reload the page, just pass the locale to momentJS.

### CODE REVIEW

File: lib/private/Settings/Personal/PersonalInfo.php [hunk 1]
```diff
@@ -258,6 +260,50 @@ private function getLanguages(IUser $user) {
...
+		$userLang = $this->config->getUserValue($uid, 'core', 'lang', $this->l10nFactory->findLanguage());
+
+		$localeData = file_get_contents(__DIR__ . '/locales.json');
+		$localeCodes = json_decode($localeData, true);
+
+		$userLocale = array_filter($localeCodes, function($value) use ($userLocaleString) {
+			return $userLocaleString === $value['code'];
+		});
+
+		if (count($userLocale) > 0)
```
Comment (nickvergessen) [inline lib/private/Settings/Personal/PersonalInfo.php]: `!empty()`

File: settings/templates/settings/personal/personal.info.php [hunk 2]
```diff
@@ -314,3 +315,29 @@
 	</a>
 </form>
 <?php } ?>
+
+<form id="locale" class="section">
```
Comment (nickvergessen) [inline settings/templates/settings/personal/personal.info.php]: Move above the closing bracket? You don't need the dropdown when you can't save it.

File: settings/templates/settings/personal/personal.info.php [hunk 3]
```diff
@@ -314,3 +315,29 @@
...
+		<option value="<?php p($_['activelocale']['code']);?>">
+			<?php p($_['activelocale']['name']);?>
+		</option>
+		<?php foreach($_['locales'] as $locale):?>
+			<option value="<?php p($locale['code']);?>">
+				<?php p($locale['name']);?>
+			</option>
+		<?php endforeach;?>
+	</select>
+</form>
```
Comment (nickvergessen) [inline settings/templates/settings/personal/personal.info.php]: New line

File: lib/private/L10N/Factory.php [hunk 4]
```diff
@@ -111,7 +112,7 @@ public function get($app, $lang = null) {
 
 		if (!isset($this->instances[$lang][$app])) {
 			$this->instances[$lang][$app] = new L10N(
-				$this, $app, $lang,
+				$this, $app, $lang, $locale,
```
Comment (nickvergessen) [inline lib/private/L10N/Factory.php]: locale is not filled with the users config?!

**Review by nickvergessen — DISMISSED** [review]

**Comment by nickvergessen** [comment]
> Migrate the core apps to use this

I disagree. This should happen automatically. It should automatically pick up the user's config and use that localle if none was given explicit, just like with the language. The cases where we set the language explicit are a different story of course, but there shouldn't be a lot.

**Comment by georgehrke** [comment]
>  Migrate the core apps to use this

Ideally core apps would just use momentJS and we set moment.locale($locale) in core.js

**Comment by tcitworld** [comment]
> Ideally core apps would just use momentJS and we set moment.locale($locale) in core.js

Done.

**Comment by georgehrke** [comment]
There is too much space between Language and Locale: 

nextcloud chromium today at 3 39 49 pm

Additionally I would like to show some examples for the selected locale.

**Comment by georgehrke** [comment]
AFAIK you can simply update the locale with `moment.locale(newLocale)`. There is no need to do a full reload.

**Comment by georgehrke** [comment]
How did you acquire the data for `lib/private/L10N/locales.json`?

Is there an automated way to update it?

**Comment by MorrisJobke** [comment]
@tcitworld Could you please resolve the conflict? Thanks 👍

**Comment by tcitworld** [comment]
> There is too much space between Language and Locale:

Yeah, don't really know why.

> Additionally I would like to show some examples for the selected locale.

Meaning ?

> How did you acquire the data for lib/private/L10N/locales.json?

Originally stolen from https://gist.github.com/jacobbubu/1836273, but I've found newer versions at https://gist.github.com/jasef/337431c43c3addb2cbd5eb215b376179 and https://gist.github.com/ndbroadbent/588fefab8e0f1b459fcec8181b41b39c

I guess it could be retrieved from elsewhere (for instance Debian has a languages code list at `cat /usr/share/i18n/SUPPORTED`), but I didn't found any perfect up-to-date list.

>  Could you please resolve the conflict? Thanks

Fixed.

**Comment by georgehrke** [comment]
> Meaning ?

I thought about something like this: 
dedicated locale-region option issue 1781 nextcloud-server safari today at 2 07 48 pm

**Comment by doodhout** [comment]
Any progress on this?

Currently biding my time using en_GB as a language+locale setting, but having the option to choose both would be nice.

**Comment by georgehrke** [comment]
@tcitworld What's the current state? :) 
Do you need any help?

**Comment by georgehrke** [comment]
> @tcitworld What's the current state? :)
Do you need any help?

> How did you acquire the data for `lib/private/L10N/locales.json`?
Is there an automated way to update it? 

ping :)

File: lib/private/Settings/Personal/PersonalInfo.php [hunk 5]
```diff
@@ -258,6 +260,41 @@ private function getLanguages(IUser $user) {
...
+
+		$localesForLanguage = array_filter($localeCodes, function($localeCode) use ($userLang) {
+			return 0 === strpos($localeCode['code'], $userLang);
+		});
+
+		return [
+			'activelocaleLang' => $userLocaleString,
+			'activelocale' => $userLocale,
+			'locales' => $localeCodes,
+			'localesforlanguage' => $localesForLanguage,
```
Comment (nickvergessen) [inline lib/private/Settings/Personal/PersonalInfo.php]: Want to use correct CamelCasing?

File: lib/private/L10N/L10N.php [hunk 6]
```diff
@@ -76,6 +82,15 @@ public function getLanguageCode() {
 	}
 
 	/**
+	 * The code (en_US, fr_CA, ...) of the locale that is used for this instance
+	 *
+	 * @return string locale
+	 */
+	public function getLocaleCode() {
+		return $this->locale;
```
Comment (nickvergessen) [inline lib/private/L10N/L10N.php]: missing the fallback for null which you do in  `l()` ? Should we do this here and use the method then below?

**Comment by georgehrke** [comment]
Looking at moment.js, if offers way fewer locales than the locale.json.
(`moment.locales()`)

Unfortunately moment.js doesn't offer readable names for each locale.
Hence I would abandon our current locale.json und generate our own based on the locales actually supported by moment.js

**Comment by tcitworld** [comment]
In that case we could also try to translate the locale names in their own languages, like the language field does.

**Comment by georgehrke** [comment]
We should be able to parse it from the moment-with-locales.js.
Every local contains this header: https://github.com/moment/moment/blob/develop/locale/de.js#L2

Unfortunately there is no localized version of the name.

**Comment by MorrisJobke** [comment]
Sorry - we need to move this to 14 :/

**Comment by skjnldsv** [comment]
It could be great to have a summary to what choosing this locale will really change, like the date, a file size example, the am/pm... etc etc :)

**Comment by biva** [comment]
Hello, any news on this PR? Any chance to change the first day of the week soon? Thank you for your work!

**Comment by biva** [comment]
Hello, any chance to see this in NC14? Thanks again for this work!

**Comment by georgehrke** [comment]
There are only a couple of missing things here:
- also set local in `js.js`
- tests
- decide where list of locales should go
- rebase

> decide where list of locales should go:

Right now they are in `lib/private/L10N/locales.json`. 
@MorrisJobke Any opinion on the location of that file?

**Comment by MorrisJobke** [comment]
> Right now they are in lib/private/L10N/locales.json.
> @MorrisJobke Any opinion on the location of that file?

we could also put it in the resources folder where also the mimetype mapping and aliases are.

**Comment by georgehrke** [comment]
Rebased, I will move the locales.json, add moment to js.js and look into the tests.

@skjnldsv Could you help out with the design / layouting in the personal settings?
it looks a bit odd:
bda39ed7-6d71-4018-982d-cbaa8194cb12
fa480bc0-cd82-409e-8b5a-2282ca5e32d3

**Comment by skjnldsv** [comment]
> @skjnldsv Could you help out with the design / layouting in the personal settings?

Of course :)

**Comment by MorrisJobke** [comment]
Could you incorporate the "enforce-language" setting into this option? Just do not show the option if the language is enforced (then only admins can change it).

File: lib/private/L10N/L10N.php (hunk 6 above)
Comment (georgehrke) [reply lib/private/L10N/L10N.php]: Handled up here: https://github.com/nextcloud/server/pull/5623/files#diff-0adb6edb83bb2d21c924c900685ed501R128

File: lib/private/L10N/Factory.php [hunk 7]
```diff
@@ -207,6 +207,11 @@ public function findLanguage($app = null) {
 	 * @return null|string
 	 */
 	public function findLocale($lang = null) {
+		$forceLocale = $this->config->getSystemValue('force_locale', false);
```
Comment (MorrisJobke) [inline lib/private/L10N/Factory.php]: Please also document it in the sample config.

Comment (georgehrke) [reply lib/private/L10N/Factory.php]: That's already the case :)
https://github.com/nextcloud/server/pull/5623/files#diff-2f769bc01ffefdf14eb04404218e7582

**Comment by georgehrke** [comment]
It's done from a coding point of view imho. Just the layouting fixes :)

File: lib/public/IL10N.php [hunk 8]
```diff
@@ -107,4 +108,12 @@ public function l(string $type, $data, array $options = []);
 	 * @since 7.0.0
 	 */
 	public function getLanguageCode(): string ;
+
+	/**
+	 * * The code (en_US, fr_CA, ...) of the locale that is used for this IL10N object
+	 *
+	 * @return string locale
+	 * @since 13.0.0
```
Comment (MorrisJobke) [inline lib/public/IL10N.php]: 14 ;)

File: lib/public/L10N/IFactory.php [hunk 9]
```diff
@@ -44,6 +44,13 @@ public function get($app, $lang = null);
 	 */
 	public function findLanguage($app = null);
 
+	/**
+	 * @param string|null $lang user language as default locale
+	 * @return string locale If nothing works it returns 'en_US'
+	 * @since 13.0.0
```
Comment (MorrisJobke) [inline lib/public/L10N/IFactory.php]: 14 ;)

File: lib/public/L10N/IFactory.php [hunk 10]
```diff
@@ -53,11 +60,33 @@ public function findLanguage($app = null);
...
+	 */
+	public function localeExists($locale);
+
+	/**
+	 * Creates a function from the plural string
+	 *
+	 * @param string $string
+	 * @return string Unique function name
+	 * @since 9.0.0
```
Comment (MorrisJobke) [inline lib/public/L10N/IFactory.php]: And this especially ;)

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍  (beside the little nitpicks the code looks good as well)

**Review by skjnldsv — APPROVED** [review]

**Comment by georgehrke** [comment]
Fixed the docblock since mistakes

**Comment by danger89** [comment]
Does this also introduces a way for an admin to set the DEFAULT locale/region? Let's say you create a new user.

**Comment by danger89** [comment]
Ps. Also update the docs? https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/language_configuration.html?highlight=language#

**Comment by tcitworld** [comment]
@danger89 
> I see, this is not an GUI option set. I think the only way is to set defaults within the config file. Correct?

Yup, just like for the language.

**Comment by skjnldsv** [comment]
@tcitworld can you provide a pr for the docs ? :)

**Comment by tcitworld** [comment]
@skjnldsv https://github.com/nextcloud/documentation/pull/1051

**Pull request merged**